File: src/python/main/helper/rich_json_logger.py
import sys
import time
import logging
from ..other.rich_json_configuration import _RICH_JSON_CONFIG

logging.basicConfig(
    level=logging.DEBUG,
    stream=sys.stdout,
    format='[%(levelname)s] %(message)s',
    force=True
)

class RichJsonLogger:
    logger = logging.getLogger("RichJSON")

    def __init__(self, label):
        self._label = label
        self._padding = ""
        self.start = None
        self.end = None

    def info(self, message):
        if _RICH_JSON_CONFIG["info_enabled"]:
            RichJsonLogger.logger.info(self._build_message(message))

    def _build_message(self, message):
        message = self._label + " " + message
        return message

    def debug(self, message):
        if _RICH_JSON_CONFIG["debug_enabled"]:
            RichJsonLogger.logger.debug(self._build_message(message))

    def error(self, message):
        RichJsonLogger.logger.error(self._build_message(message))

    def group_start(self):
        if _RICH_JSON_CONFIG["info_enabled"] or _RICH_JSON_CONFIG["debug_enabled"]:
            self._padding += "  "
            logging.basicConfig(
                level=logging.DEBUG,
                stream=sys.stdout,
                format=self._padding + '[%(levelname)s] %(message)s',
                force=True
            )

    def group_end(self):
        if _RICH_JSON_CONFIG["info_enabled"] or _RICH_JSON_CONFIG["debug_enabled"]:
            self._padding = self._padding[:-2]
            logging.basicConfig(
                level=logging.DEBUG,
                stream=sys.stdout,
                format=self._padding + '[%(levelname)s] %(message)s',
                force=True
            )

    def group_end_all(self):
        if _RICH_JSON_CONFIG["info_enabled"] or _RICH_JSON_CONFIG["debug_enabled"]:
            logging.basicConfig(
                level=logging.DEBUG,
                stream=sys.stdout,
                format='[%(levelname)s] %(message)s',
                force=True
            )

    def time_start(self):
        if _RICH_JSON_CONFIG["debug_enabled"]:
            self.start = time.perf_counter_ns()

    def time_end(self):
        if _RICH_JSON_CONFIG["debug_enabled"]:
            self.end = time.perf_counter_ns()
            self.debug(f"{self.end - self.start} ns")

File: src/js/main/helper/RichJsonLogger.js
import {__RICH_JSON_CONFIG as _RICH_JSON_CONFIG} from "../other/RichJsonConfiguration.js";

export class RichJsonLogger {
    groupLevel = 0;

    constructor(label) {
        this._label = label;
        this._padding = "";
        this.start = null;
        this.end = null;
    }

    /**
     * Internal helper to format the message with level, label, and padding
     */
    _formatMessage(message) {
        return `${this._padding}${this._label} ${message}`;
    }

    info(message) {
        if (_RICH_JSON_CONFIG.infoEnabled) {
            console.info(this._formatMessage(message));
        }
    }

    debug(message) {
        if (_RICH_JSON_CONFIG.debugEnabled) {
            console.debug(this._formatMessage(message));
        }
    }

    error(message) {
        console.error(this._formatMessage(message));
    }

    groupStart(message) {
        if (_RICH_JSON_CONFIG.infoEnabled) {
            this.groupLevel++;
            console.group(this._formatMessage(message));
        }
    }

    groupStartDebug(message) {
        if (_RICH_JSON_CONFIG.debugEnabled) {
            this.groupLevel++;
            console.group(this._formatMessage(message));
        }
    }

    groupEnd() {
        if (_RICH_JSON_CONFIG.infoEnabled) {
            this.groupLevel--;
            console.groupEnd();
        }
    }

    groupEndDebug() {
        if (_RICH_JSON_CONFIG.debugEnabled) {
            this.groupLevel--;
            console.groupEnd();
        }
    }

    groupEndAll() {
        if (_RICH_JSON_CONFIG.infoEnabled || _RICH_JSON_CONFIG.debugEnabled) {
            for (let i = 0; i < this.groupLevel; i++) {
                console.groupEnd();
            }
        }
    }

    timeStart() {
        if (_RICH_JSON_CONFIG.debugEnabled) {
            this.start = performance.now();
        }
    }

    timeEnd() {
        if (_RICH_JSON_CONFIG.debugEnabled && this.start !== null) {
            this.end = performance.now();
            const durationNs = Math.round((this.end - this.start) * 1_000_000);
            this.debug(`${durationNs} ns`);
        }
    }
}